# Estatística & Probabilidade - 2023/02

## Projeto 01 - Calculando probabilidades com simulação estocástica

### Hugo Tremonte de Carvalho

#### hugo@dme.ufrj.br

#### Google Classroom: `oug6dox`

# ⚠️ INSTRUÇÕES ⚠️ LEIAM ATENTAMENTE! ⚠️

* A data limite de entrega da avaliação é sábado 01/10/2023 às 23h59. Avaliações entregues após esse prazo serão desconsideradas.

* A entrega deve ser feita exclusivamente através do Google Classroom, usando a linguagem Python dentro de um notebook do Google Colaboratory, a ser anexado na atividade da seguinte forma:
  - Dentro da turma no Google Classroom, clique em "Atividades"
            
  - No tema "Projetos", clique em "Projeto 02" e em seguida em "Ver atividade"
            
  - No canto superior direito, clique em "+ Adicionar ou criar" e em seguida selecione em "Arquivo" ou "Google Drive", se optar fazer upload do notebook ou selecionar um arquivo de seu Google Drive, respectivamente.
            
  - O título do notebook deverá ser, obrigatoriamente, "SEU NOME - Projeto 01".
            
  - Para entregar o projeto, abra a atividade que o contém (conforme procedimento explicado no ponto acima), e no canto superior direito clique em "ENTREGAR" (o texto estará em uma caixa cinza, mas ele é clicável).
            
  - Na tela a seguir clique em "Entregar" para confirmar a entrega do trabalho.
            
  - Se você se arrependeu e quiser fazer outra entrega, basta abrir a atividade, no canto superior direito clique em "Cancelar envio" e confirme clicando novamente em "Cancelar envio" na janela que irá abrir. Você pode fazer isso quantas vezes quiser até o prazo determinado.
            
  - As explicações que você precisará fazer devem ser feitas em células de texto em seu notebook. Somente comentários pequenos devem ser feitos no código.
            
  - Caso seja necessário, equações podem ser introduzidas em $\LaTeX$ nas células de texto.
            
  - Figuras também podem ser incluídas em células de texto, tanto através de upload quanto simplesmente arrastando-a para alguma célula de texto. Nesse caso, sempre informe a fonte de onde a figura foi retirada, caso não tenha sido gerada por você.

* O projeto poderá ser feito **até** em dupla: quem quer fazer sozinho/a faz, mas quem quiser fazer em dupla também tem esse direito. Porém, vamos evitar problemas de "eu fiz tudo e ele/a não fez nada"! Infelizmente não tenho como controlar isso, e após o projeto entregue não adianta "eu fiz tudo sozinho/a", pois eu não estava lá para ver! Tentem se resolver entre si, mas em último caso posso auxiliar com uma mediação.

* No caso de trabalho em dupla, apenas **uma** pessoa da dupla deverá fazer o envio do trabalho, claro, deixando explícito quem é o/a parceiro/a.

* A troca de conhecimento na realização da avaliação é permitida e encorajada: ciência se faz com colaboração, e devemos seguir esse espírito aqui. Porém, cada dupla (ou indivíduo) deverá ter a sua própria resolução, e cópias ou outras ilegalidades serão severamente punidas com a anulação da avaliação para o(s) aluno(s) suspeito(s).

* Todos os passos de sua resolução devem ser devidamente justificados.

* Ao entregar essa avaliação, você afirma ter lido e estar de acordo com essas regras, comprometendo-se a cumpri-las.

# O projeto

O objetivo da atividade é que vocês calculem algumas probabilidades através de simulação estocástica, ou seja, através de um grande número de sorteios aleatórios devidamente projetados para calcular a probabilidade desejada. Algumas dessas probabilidades podem ser calculadas analiticamente com certa facilidade, mas em outros pode ser **bem** difícil (ou mesmo impossível). Dessa forma, o cálculo de probabilidades através de simulação estocástica pode ser tanto uma ferramenta fundamental para resolver algum problema ou para te dar alguma intuição, te guiando para o resultado analítico. Em alguns momentos vocês deverão justficar teoricamente os resultados obtidos, pois senão, não tem graça! 😁

## Questão 1: Calculando $\pi$ jogando dardos

a) Suponha que num jogo de dardos você tem, pendurado em uma parede, o alvo (círculo de raio 1) devidamente emoldurado (inscrito em um quadrado de lado 2 - as unidades não importam aqui). Porém, suas habilidades ao arremessar um dardo em direção ao alvo são péssimas, e você pode atingir qualquer ponto dentro do quadrado uniformemente (note que, forças ocultas e inexplicáveis do Universo fazem com que seja impossível você acertar *fora* da moldura). Através de um cálculo analítico mostre como que a probabilidade de você acertar o alvo se relaciona com o número $\pi$.

Abaixo se encontra uma representação gráfica do círculo de raio 1 inscrito no quadrado de lado 2.

<img src=cir.png>

Para calcular a probabilidade de acertar o alvo, é necessário utilizar a fórmula abaixo:                          

P(A) = número de elementos em A = $\frac{eventos favoraveis}{eventos\; totais}$

Podemos considerar os eventos totais, acertar qualquer lugar dentro do quadrado, sendo assim a área do quadrado.

eventos totais = área do quadrado = $2^2$ = 4

Agora os eventos favoráveis é acertar qualquer lugar dentro do cíeculo, sendo assim a área do círculo.

eventos favoráveis = área do círculo = $\pi*1^2$ = $\pi$

Dessa forma, a probabilidade de você acertar o alvo é $\frac{\pi}{4}$.

Logo a probabilidade de acretar o alvo é um quarto do valor de $\pi$ neste caso.

Ainda, podemos relacionar a probabilidade com a área do quadrado, sendo L o lado, a área do quadrado é igual a $L^2$. E o raio do círculo sendo R, temos área do círculo = $\pi*R^2$

Então a probabilidade sempre será eventos favoráveis = área do círculo = $\pi*R^2$ sobre eventos totais = área do quadrado = $L^2$ 
sendo $\frac{\pi*R^2}{L^2}$.

b) Utilizando geração de números aleatórios, faça uma simulação para estimar o valor de $\pi$ através do procedimento acima. Ou seja, simule que dardos são arremessados uniformemente no quadrado e utilize a interpretação frequentista da probabilidade para aproximar o valor de $\pi$. Note que você precisará de "muitas" tentativas para ter uma boa estimativa. O quão grande você acha que "muitas" deve ser?

Para construir a simulação precisamos modelar o problema, podemos posicionar o quadrado em um plano cartesiano, de forma que o centro do circulo inscrito esteja na posição [0,0], como mostra na figura abaixo

<img src=plano-cartesiano.png width='400px'>

Observando a figura, podemos concluir que para que um ponto esteja presente no quadrado, as coordenadas $x$ e $y$ devem variar de -1 a 1. Desta forma uma maneira uma forma de simular a jogada do dardo, é sotear uma coordenada $x$ e $y$ seguindo os segintes intervalos $-1 \leq x \leq 1$ e $-1 \leq y \leq 1$.

Sendo assim falta só saber se a coordenada sorteada está dentro do círculo, para isso vamos usar a equação do círculo.

$$x^2 + y^2 = 1$$

A equação verifica se o ponto está localizado na extremidade do circulo, para considerar pontos no interior do círculo, iremos trocar o igual por um menor igual.

$$x^2 + y^2 \leq 1$$

Com todas essas informações consideradas, podemos montar a nossa simulação

A nossa simulação irá simular a jogada de um dardo várias vezes, e contabilizar as vezes que o alvo foi acertado, e depois dividir pelo número de lancamentos, para ter o resultado da probabilidade de acertar o alvo.

In [1]:
import random

def lançar_dardos(n):
    contador_Acertos = 0
    for _ in range(n):
        x = random.uniform(-1, 1)
        y = random.uniform(-1, 1)
        if((x**2 + y**2)<=1):
            contador_Acertos = contador_Acertos + 1
    return contador_Acertos/n


numero_de_lancamentos = 100000000
resultado = lançar_dardos(numero_de_lancamentos)
print("Probabilidade de acertar o alvo: ", resultado)


Probabilidade de acertar o alvo:  0.78537951


Pelo resultado obtido no item a, a probabilidade de você acertar o alvo é $\frac{\pi}{4}$. Dessa forma podemos igualar isso ao resultado obtido na simulação e assim ter uma estimativa do valor de $\pi$.

$$\frac{\pi}{4} \approx 0.78548015$$

$$\pi \approx 0.78548015*4$$

$$\pi \approx 3.1419206$$

## Questão 2: O problema de Monty Hall

Em um famoso programa de televisão da década de 70 nos Estados Unidos, Monty Hall, o apresentador, apresentava três portas aos concorrentes. Atrás de uma delas estava um prêmio e, atrás das outras duas não havia nada. A interação se dava em três etapas:

* Na 1a. etapa o concorrente escolhe uma das três portas (que ainda não é aberta);
* Na 2a. etapa, Monty abre uma das outras duas portas que o concorrente não escolheu, revelando que o prêmio não se encontra nessa porta e revelando uma porta vazia (note que o apresentador **sempre** irá abrir uma porta que está vazia, e **nunca** a porta com o prêmio);
* Na 3a. etapa Monty pergunta ao concorrente se quer decidir permanecer com a porta que escolheu no início do jogo ou se ele pretende mudar para a outra porta que ainda está fechada para então a abrir. Agora, com duas portas apenas para escolher — pois uma delas já se viu, na 2a. etapa, que não tinha o prêmio — e sabendo que o prêmio está atrás de uma das restantes duas, o concorrente tem que tomar a decisão.

O objetivo dessa questão é estudar qual é a estratégia mais lógica: ficar com a porta escolhida inicialmente ou mudar de porta? Do ponto de vista probablístico, com qual das duas portas ainda fechadas o concorrente tem mais probabilidades de ganhar?

a) Primeiramente, pensando do ponto de vista do participante, argumente se é mais razoável que ele tome sua decisão pensando do ponto de vista da interpretação frequentista ou subjetiva do conceito de probabilidade. Claro, justifique sua escolha!

Obs.: Esse item é engraçado pois uma vez que o cálculo da probabilidade independe da interpretação escolhida, qualquer resposta é válida, do ponto de vista formal. Porém, a graça aqui é *como* você vai fazer a sua justificativa para a interpretação escolhida!

A partir de uma perspectiva subjetiva, a probabilidade é interpretada como uma representação da incerteza pessoal que um indivíduo tem em relação a um evento. Nessa abordagem subjetiva, levo em conta a atualização das minhas crenças com base nas informações disponíveis e na minha própria incerteza. Isso elimina a necessidade de realizar inúmeros experimentos para calcular a probabilidade de acerto, pois as minhas crenças pessoais refletem a minha avaliação subjetiva da situação.-

b) Talvez você já saiba a resposta para esse problema, uma vez que ele é bem famoso, mas não tem problema! Faça simulações para encontrar qual é o melhor cenário para o participante: manter sua escolha inicial ou mudar de porta. Justifique, **intuitivamente** (ou seja, **sem fazer contas**) porque essa escolha é razoável.

-- sua resolução aqui --

In [74]:
import random

def simulador_portas(n):
    contador_troca = 0
    contador_nao_troca = 0
    
    for _ in range(n):
        # Colocar o prêmio em uma das portas
        premio = random.randint(0, 2)
        # Porta escolhida
        porta_escolhida = random.randint(0, 2)

        # Monty apresenta uma porta vazia
        porta_vazia = list(set([0, 1, 2]) - set([premio, porta_escolhida]))
        porta_de_monty = random.choice(porta_vazia)

        # Determinar a outra porta fechada que não foi escolhida
        porta_fechada_nao_escolhida = list(set([0, 1, 2]) - set([porta_escolhida, porta_de_monty]))[0]
        
        # Verificar qual das portas não abertas contém o prêmio
        if porta_fechada_nao_escolhida == premio:
            # Contar as vezes que a porta não escolhida contém o prêmio
            contador_troca += 1
        else:
            # Contar as vezes que a porta escolhida contém o prêmio
            contador_nao_troca += 1
    return (contador_nao_troca / n, contador_troca / n)        

probabilidade_de_acerto_ao_nao_trocar, probabilidade_de_acerto_ao_trocar = simulador_portas(10 ** 6)

print('Probabilidade de ganhar o prêmio ao trocar de porta:', probabilidade_de_acerto_ao_trocar)
print('Probabilidade de ganhar o prêmio ao não trocar de porta:', probabilidade_de_acerto_ao_nao_trocar)


Probabilidade de ganhar o prêmio ao trocar de porta: 0.667093
Probabilidade de ganhar o prêmio ao não trocar de porta: 0.332907


c) Agora sim, justifique fazendo as contas a probabilidade encontrada no item b). Caso você pesquise por aí, você vai encontrar algumas soluções utilizando o Teorema de Bayes. Como nós não vimos ainda tal conteúdo, tente fazer somente com probabilidade condicional. Juro que é possível, e vai ser legal para treinar o conceito recém-aprendido.

Primeiramente, começaremos com a escolha de porta do participante, temos 3 portas fechadas então o número de eventos totais é igual a 3. O participante escolhe uma porta, sendo o número de eventos favoráveis igual a 1. Então, a probabilidade do participantes escolher a porta premiada em sua primeira escolha é igual a 1/3. Esse valor corresponde ao visto na simulação onde a Probabilidade de ganhar o prêmio ao não trocar de porta é aproximadamente 0.332907. 

Agora, pensemos na probabilidade de ganhar o prêmio ao trocar de porta. Nessa, temos uma mudança significativa de cenário, pois uma das portas que não contém o prêmio já foi aberta. Então queremos encontrar o evento A = a segunda porta escolhida (diferente da primeira) é a premiada, dado o evento B = uma porta que não contém o prêmio ser aberta. 
Analisemos:

Para a primeira porta escolhida estar com prêmio a probabilidade é 1/3, como mostrado acima.
Para a primeira porta escolhida não ter prêmio a probabilidade é 2/3, pois é o complementar da anterior (1 - 1/3).

Para achar a probabilidade de ganhar o prêmio ao trocar de porta, vamos trabalhar com os dois casos acima.
Se a primeira porta escolhida tem prêmio, então a probabilidade de a segunda porta escolhida ter prêmio é igual a 0.

Agora, tendo que a primeira porta escolhida não tem prêmio, nesse caso se o participante manter a porta escolhida primeiramente as chances de ganhar são iguais a 0. Então, tendo em vista que uma porta que não tinha prêmio foi aberta, a chance dele ganhar se trocar a porta vai ser complementar a probabilidade de não trocar a porta, pois só há uma opção de porta. Sendo 2/3 a probabilidade para este caso de a primeira porta escolhida nao ter o prêmio, então o complementar aqui será 2/3 - 0 = 2/3 de chances de a segunda porta escolhida ter o prêmio. Esse valor também foi o encontrado na simulação do item b) que resultou em aproximadamente 0.667093


Também, a seguir, uma representação gráfica das condicionais relatadas:

<img src=probestgrafico.png>

## Questão 3: Jogo da velha aleatório

Agora, vamos tentar estimar uma probabilidade cujo cálculo analítico é bem mais complexo. Suponha que dois jogadores joguem jogo da velha porém com uma estratégia um pouco exótica: cada jogador escolhe a casa onde irá marcar $\times$ ou $\circ$ uniformemente dentre as casas vazias. Exemplificando, o primeiro jogador marca $\times$ em uma das 9 casas com igual probabilidade; o segundo marca $\circ$ em uma das 8 casas restantes com igual probabilidade; e assim sucessivamente até algum jogador ganhar (linha, coluna ou diagonal completa com $\times$ ou $\circ$) ou empatar (todas as células preenchidas porém sem nenhuma linha nem coluna nem diagonal completa com $\times$ ou $\circ$).

a) Encontre uma aproximação para a probabilidade do jogo **não** terminar em empate. Tente encontrar uma justificativa intuitiva para sua resposta.

Iremos fazer apresentar um codigo que irá simular as partidas de jogo da velha para obter uma estimativa da probabilidade do jogo não terminar em empate.

para poder simular a partida é necessáio estabelecer que o jogo só termina em empate se todas as células preenchidas porém sem nenhuma linha nem coluna nem diagonal completa com $\times$ ou $\circ$

In [114]:
import random
def verificar_vitoria(tabuleiro, jogador):
    # Verifica linhas e colunas
    for i in range(3):
        if (tabuleiro[i][0] == tabuleiro[i][1] == tabuleiro[i][2] == jogador ) or \
        (tabuleiro[0][i] == tabuleiro[1][i] == tabuleiro[2][i]  == jogador ):
            return True

    # Verifica diagonais
    if all(tabuleiro[i][i] == jogador for i in range(3)) or all(tabuleiro[i][2 - i] == jogador for i in range(3)):
        return True

    return False


def simular_jogo_da_velha(n):
    vitorias = 0

    for _ in range(n):
        # Inicializa o tabuleiro vazio
        tabuleiro = [[' ', ' ', ' '], [' ', ' ', ' '], [' ', ' ', ' ']]
        
        jogador = 'X'
        casas_vazias = [(i, j) for i in range(3) for j in range(3)] 

        while True:  # Número máximo de movimentos é 9
            
            if casas_vazias:
                i, j = random.choice(casas_vazias)
                tabuleiro[i][j] = jogador
                casas_vazias.remove((i,j))

                # Verifica se o jogador atual venceu
                if verificar_vitoria(tabuleiro, jogador):
                    vitorias += 1
                    break

                # Alterna para o próximo jogador
                jogador = 'O' if jogador == 'X' else 'X'
            else:
                # O jogo termina em empate
                break

    # Calcula a probabilidade de vitória do jogador X
    return vitorias/n

# Exemplo de uso
Pvitorias=simular_jogo_da_velha(10**5)  
print('Probabilidade do jogo não terminar empatado')
print(Pvitorias)

Probabilidade do jogo não terminar empatado
0.87369


-- sua resolução aqui --

b) Dentre os jogos que não terminam empatados, qual jogador tem mais chances de ganhar, se ambos utilizam essa estratégia: o jogador que começa ou o segundo jogador? Tente encontrar uma justificativa intuitiva para sua resposta.

In [18]:
import random
def verificar_vitoria(tabuleiro, jogador):
    # Verifica linhas e colunas
    for i in range(3):
        if (tabuleiro[i][0] == tabuleiro[i][1] == tabuleiro[i][2] == jogador ) or \
        (tabuleiro[0][i] == tabuleiro[1][i] == tabuleiro[2][i]  == jogador ):
            return True

    # Verifica diagonais
    if all(tabuleiro[i][i] == jogador for i in range(3)) or all(tabuleiro[i][2 - i] == jogador for i in range(3)):
        return True

    return False


def simular_jogo_da_velha(n):
    vitorias = 0
    vitorias_X = 0
    vitorias_O = 0
    for _ in range(n):
        # Inicializa o tabuleiro vazio
        tabuleiro = [[' ', ' ', ' '], [' ', ' ', ' '], [' ', ' ', ' ']]
        
        jogador = 'X'
        casas_vazias = [(i, j) for i in range(3) for j in range(3)] 

        while True:  # Número máximo de movimentos é 9
            
            if casas_vazias:
                i, j = random.choice(casas_vazias)
                tabuleiro[i][j] = jogador
                casas_vazias.remove((i,j))

                # Verifica se o jogador atual venceu
                if verificar_vitoria(tabuleiro, jogador):
                    vitorias += 1
                    if(jogador == 'X'):
                        vitorias_X +=1
                    else:
                        vitorias_O +=1
                    break

                # Alterna para o próximo jogador
                jogador = 'O' if jogador == 'X' else 'X'
            else:
                # O jogo termina em empate
                break


    return vitorias_X/vitorias, vitorias_O/vitorias

# Exemplo de uso
Pjogador1, Pjogador2 =simular_jogo_da_velha(10**5)  

print('probabilidade do 1 jogador ganhar: ')
print(Pjogador1)
print('probabilidade do 2 jogador ganhar: ')
print(Pjogador2)

probabilidade do 1 jogador ganhar: 
0.6705576429618768
probabilidade do 2 jogador ganhar: 
0.3294423570381232


c) Responda às mesmas perguntas dos itens a) e b), porém para um *jogo da velha tridimensional*: ou seja, em vez do jogo acontecer em uma grade bidimensional de tamanho $3 \times 3$, agora ele acontece em uma grade tridimensional de tamanho $3 \times 3 \times 3$! 🤯 Tome cuidado pois agora há diversas formas de um jogador ganhar uma partida! Além de ser através do completamento de alguma linha, coluna ou diagonal de qualquer "sub-tabuleiro" plano do cubo $3 \times 3 \times 3$ (existem 9 desses tabuleiros), o jogador poderá ainda completar as diagonais do cubo que unem vértices opostos (4 no total).

Na estratégia de simulação, começaremos com o tabuleiro inicial e dividiremos em nove subtabuleiros, nos quais testaremos a condição de vitória, utilizando o código já apresentado. Além disso, faremos uma verificação adicional para determinar a vitória nas diagonais.

In [113]:
import random

def verificar_vitoria_2d(tabuleiro, jogador):
    # Verifica linhas e colunas
    for i in range(3):
        if (tabuleiro[i][0] == tabuleiro[i][1] == tabuleiro[i][2] == jogador ) or \
        (tabuleiro[0][i] == tabuleiro[1][i] == tabuleiro[2][i]  == jogador ):
            return True

    # Verifica diagonais
    if all(tabuleiro[i][i] == jogador for i in range(3)) or all(tabuleiro[i][2 - i] == jogador for i in range(3)):
        return True

    return False

def extrair_subtabuleiros(cubo):
    subtabuleiros = []
    subtabuleiros.append(cubo[0])
    subtabuleiros.append(cubo[1])
    subtabuleiros.append(cubo[2])
    s =[cubo[0][0], cubo[1][0], cubo[2][0]]
    subtabuleiros.append(s)
    s =[cubo[0][1], cubo[1][1], cubo[2][1]]
    subtabuleiros.append(s)
    s =[cubo[0][2], cubo[1][2], cubo[2][2]]
    subtabuleiros.append(s)
    for coluna in range(3):
        col = []
        for plano in range(3):
            col.append([linha[coluna] for linha in cubo[plano]])
        subtabuleiros.append(col)
    return subtabuleiros

def verificar_vitoria_3d(tabuleiro, jogador):
    # Verifica vitória nos "sub-tabuleiros" planos de 3x3]]
    tabuleiros = extrair_subtabuleiros(tabuleiro)

    for camada in tabuleiros:
        if verificar_vitoria_2d(camada, jogador):
            return True
    #Verificar a vitoria nas diagonais
    if (all(tabuleiro[i][i][i] == jogador for i in range(3)) or
         all(tabuleiro[i][i][2 - i] == jogador for i in range(3)) or
         all(tabuleiro[i][2 - i][i] == jogador for i in range(3)) or
         all(tabuleiro[i][2 - i][2 - i] == jogador for i in range(3))):
         return True

    return False


def simular_xadrez_3d(n):
    vitorias = 0
    vitorias_X = 0
    vitorias_O = 0

    for _ in range(n):
        # Inicializa o tabuleiro vazio
        tabuleiro = [[[' ' for _ in range(3)] for _ in range(3)] for _ in range(3)]
       
        jogador = 'X'

        casas_vazias = [(i, j, k) for i in range(3) for j in range(3) for k in range(3)]

        while True:  # Número máximo de movimentos é 27 (3x3x3)

            if casas_vazias:
                i, j, k = random.choice(casas_vazias)
                tabuleiro[i][j][k] = jogador
                casas_vazias.remove((i, j, k))

                # Verifica se o jogador atual venceu
                if verificar_vitoria_3d(tabuleiro, jogador):
                    vitorias += 1
                    if jogador == 'X':
                        vitorias_X += 1
                    else:
                        vitorias_O += 1
                    break

                # Alterna para o próximo jogador
                jogador = 'O' if jogador == 'X' else 'X'
            else:
                
                # O jogo termina em empate
                break

    # Calcula a probabilidade de vitória do Jogador 1 e do Jogador 2
    probabilidade_jogador1 = vitorias_X / vitorias
    probabilidade_jogador2 = vitorias_O / vitorias
    return probabilidade_jogador1, probabilidade_jogador2, vitorias/n

#Exemplo de uso
Pjogador1, Pjogador2, Pvitorias = simular_xadrez_3d(10**5)

print('Probabilidade do jogo não terminar empatado')
print(Pvitorias)
print('Probabilidade do Jogador 1 ganhar: ')
print(Pjogador1)
print('Probabilidade do Jogador 2 ganhar: ')
print(Pjogador2)




Probabilidade do jogo não terminar empatado
1.0
Probabilidade do Jogador 1 ganhar: 
0.5699
Probabilidade do Jogador 2 ganhar: 
0.4301
